url594c007908_hugobraga_College_tgz-pJ8-main_GraphJ8/src/algorithms/MSCBasedAlgorithm.java

```diff
@@ -0,0 +1,230 @@
+    	
+		int [][]P = uninformedAdjMatrix.getP2PPath();
+		double [][]MCPmatrix = uninformedAdjMatrix.getP2PCost();    	
+    	
+    	for (int i = 0; i < terminalsId.size(); i++) {
+    		double minCost = -1;
+    		Iterator<ElkinItem> itensIt = chosenItem.iterator();
+    		ElkinItem minItem = null;
+    		while (itensIt.hasNext()) {
+    			ElkinItem item = itensIt.next();
+    			if (item.getNode().getId() == terminalsId.get(i).intValue()) {
+    				double cost = edgeWeight.getWeight(item.getPseudo().getInformed().getId(), item.getPseudo().getUninformed().getId(), globalAdjl) + MCPmatrix[item.getPseudo().getUninformed().getId()][item.getNode().getId()];
+    				if ((cost < minCost) || (minCost == -1)) {
+    					minItem = item.copy();
+    					minCost = cost;
+    				}
+    			}
+    		}
+    		realChosenItem.add(minItem);
+    		realChosenPseudo.add(new PseudoNode(new Node(minItem.getPseudo().getInformed().getId()), new Node(minItem.getPseudo().getUninformed().getId())));
+    	}
+    	
+    	//independent part
+		Iterator<ElkinItem> itensIt = realChosenItem.iterator();
+		while (itensIt.hasNext()) {
+			ElkinItem item = itensIt.next();
+    		int prev = item.getNode().getId();//terminal
+    		
+    		int s = item.getPseudo().getUninformed().getId();
+    		if (s == prev) { //the uninformed node of the pseudo node is a terminal
+    			quasiArborescence.checkInsertEdge(item.getPseudo().getInformed().getId(), prev, edgeWeight.getWeight(item.getPseudo().getInformed().getId(), prev, globalAdjl));
+    		} else {
+        		while (P[s][prev] != s) {
+        			quasiArborescence.checkInsertEdge(P[s][prev], prev, MCPmatrix[P[s][prev]][prev]);
+        			prev = P[s][prev];
+        		}
+        		quasiArborescence.checkInsertEdge(P[s][prev], prev, MCPmatrix[P[s][prev]][prev]);
+    		}
+    		double cost = edgeWeight.getWeight(item.getPseudo().getInformed().getId(), item.getPseudo().getUninformed().getId(), globalAdjl);
+    		quasiArborescence.checkInsertEdge(item.getPseudo().getInformed().getId(), item.getPseudo().getUninformed().getId(), cost);
+    	}
+    }
+    
+	public ArrayList<Integer> applyDistDomProcedure(HashSet v1Set, ArrayList<ElkinItem> v2Set) {
+    	//finding out dominating set D
+		ArrayList<Integer> uncovered = new ArrayList<Integer>();
+		ArrayList<Group> groups = setMSCGroups(v1Set, uncovered, true);		
+    	return procDistDom(groups, v2Set);
+	}    
+	
+	//uninformedTerminals - contera o ultimo subconjunto de Next que nao tinha sido coberto anteriromente. Ou seja, contera os elementos 
+	//retornados pela ultima chamada desta funcao
+	//vId - id de um no marcado do conjunto I com menor grau que foi selecionado 
+	//v1Set - este conjunto contera, a priori, unmarked nos de I e incrementalmente ira englobando nos marked de I
+    protected ArrayList<Integer> setMSCGroupParametrized(HashSet v1Set, Group group, int uId, ArrayList<Integer> uninformedTer, ArrayList<Double> uninformedTerSpanQuant, boolean extraCostForSource) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 12. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.